aten/src/ATen/ThreadLocalState.h

```diff
@@ -87,12 +87,12 @@ class TORCH_API ThreadLocalStateGuard {
 };
 
 template <typename T>
-std::function<T(void)> wrapPropagateTLSState(
-    std::function<T(void)> callback) {
-  return [tls_state = ThreadLocalState(), callback = std::move(callback)]() {
+auto wrapPropagateTLSState(T callback) {
```

Comment:
Would I be correct if I assume we use `auto` here to avoid passing return type to the template?

Reply:
We need `auto` here for the return type because the return value is a lambda, which has an "anonymous" type. In fact, the use of `auto` here is almost identical to the use of `auto` in expressions such as:
```
auto cb = [&](Foo f) { bar(f); };
```

Remember that lambdas, in fact, are just "syntactic sugar" and that under the hood a lambda defines a class (with a random unique name, in a way) in the local scope. The fields of this class are the objects that are captured in the lambda's closure, and the `operator()` of this class is the body of the lambda. Thus `auto` will correspond to this class. Because the class has no public name, we cannot refer to it directly, and `auto` or `decltype` are our only options.

We could use `std::function` instead of `auto` but this is _not_ equivalent, as it would mean that instead of returning the raw lambda as-is, we'd wrap it in some wrapper, possibly allocating heap memory, and moving data around, etc.